examples/sampling/graphbolt/temporal_link_prediction.py

```diff
@@ -0,0 +1,367 @@
+            if step + 1 == args.early_stop:
+                # Early stopping requires a new dataloader to reset its state.
+                dataloader = create_dataloader(args, graph, features, train_set)
+                break
+
+        end_epoch_time = time.time()
+        print(
+            f"Epoch {epoch:05d} | "
+            f"Loss {(total_loss) / (step + 1):.4f} | "
+            f"Time {(end_epoch_time - start_epoch_time):.4f} s"
+        )
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description="diginetica-r2ne (GraphBolt)")
+    parser.add_argument("--epochs", type=int, default=10)
+    parser.add_argument("--lr", type=float, default=0.0005)
+    parser.add_argument("--neg-ratio", type=int, default=1)
+    parser.add_argument("--train-batch-size", type=int, default=1024)
+    parser.add_argument("--eval-batch-size", type=int, default=1024)
+    parser.add_argument("--num-workers", type=int, default=0)
+    parser.add_argument(
+        "--dataset",
+        default="diginetica-r2ne",
+        choices=["diginetica-r2ne"],
+        help="Dataset.",
+    )
+    parser.add_argument(
+        "--early-stop",
+        type=int,
+        default=0,
+        help="0 means no early stop, otherwise stop at the input-th step",
+    )
+    parser.add_argument(
+        "--fanout",
+        type=str,
+        default="20,20",
+        help="Fan-out of neighbor sampling. Default: 20, 20",
+    )
+    parser.add_argument(
+        "--exclude-edges",
+        type=int,
+        default=1,
+        help="Whether to exclude reverse edges during sampling. Default: 1",
+    )
+    parser.add_argument(
+        "--mode",
+        default="cpu-cuda",
+        choices=["cpu-cpu", "cpu-cuda"],
+        help="Dataset storage placement and Train device: 'cpu' for CPU and RAM,"
+        " 'pinned' for pinned memory in RAM, 'cuda' for GPU and GPU memory.",
+    )
+    return parser.parse_args()
+
+
+def download_datasets(name, root="datasets"):
+    url = "https://dgl-data.s3-accelerate.amazonaws.com/dataset/"
+    dataset_dir = os.path.join(root, name)
+    if not os.path.exists(dataset_dir):
+        url += name + ".zip"
+        os.makedirs(root, exist_ok=True)
+        zip_file_path = os.path.join(root, name + ".zip")
+        download(url, path=zip_file_path)
+        extract_archive(zip_file_path, root, overwrite=True)
+        os.remove(zip_file_path)
+    return dataset_dir
+
+
+def main(args):
+    if not torch.cuda.is_available():
+        args.mode = "cpu-cpu"
+    print(f"Training in {args.mode} mode.")
+    args.storage_device, args.device = args.mode.split("-")
+    args.device = torch.device(args.device)
+
+    # Load and preprocess dataset.
+    print("Loading data")
+    dataset_path = download_datasets(args.dataset)
+    dataset = gb.OnDiskDataset(dataset_path).load()
+
+    # Move the dataset to the selected storage.
+    graph = dataset.graph.to(args.storage_device)
+    features = dataset.feature.to(args.storage_device)
```

Comment:
Let's pin the features when there is a gpu. Together with moving copy_to before fetch feature, can improve performance a lot.

Reply:
You can check the PyG advanced example for the different placement of graph, features etc.

Reply:
See above.